### PR TITLE
Fix cycle dependency issue (UIContext/CppGen) is incomplete in binding.h

### PR DESCRIPTION
**Pull request by mzccn**
Previous version was not able to be compiled under clang-15 (apple version)

### CODE REVIEW

File: src/binding.h [hunk 1]
```diff
@@ -177,19 +177,8 @@ struct field_ref : property_base
...
-		const auto* var = ctx.codeGen->GetVar(str);
-		if (!var)
-			return {};
-		T val;
-		std::istringstream is(var->init);
-		if (is >> val)
-			return val;
-		return {};
-	}
+	T eval(const UIContext& ctx) const;
```
Comment (mzccn) [inline src/binding.h:180]: This is the place clang start complaining, I move to other file 'src/binding_field.h' to break the dependency

File: src/node.cpp [hunk 2]
```diff
@@ -1172,7 +1173,7 @@ void TopWindow::TreeUI(UIContext& ctx)
 			ImGui::PopStyleColor();
 		bool clicked = ImGui::IsItemClicked();
 		ImGui::SameLine();
-		ImGui::TextDisabled(NAMES[kind]);
+		ImGui::TextDisabled("%s", NAMES[kind]);
```
Comment (mzccn) [inline src/node.cpp:1176]: get complained as well, better fixing it

File: src/node.cpp [hunk 3]
```diff
@@ -1607,8 +1608,8 @@ void Widget::Draw(UIContext& ctx)
 		if (ImGui::IsMouseDragging(ImGuiMouseButton_Left))
 		{
 			ImVec2 delta = ImGui::GetMouseDragDelta();
-			if (std::signbit(pos_x + delta.x) == std::signbit(pos_x) &&
-				std::signbit(pos_y + delta.y) == std::signbit(pos_y))
+			if (std::signbit(pos_x + delta.x) == std::signbit(float(pos_x)) &&
```
Comment (mzccn) [inline src/node.cpp:1611]: explicit conversion to float as not all compiler can infer type with ease

**Review by mzccn — COMMENTED** [review]
Previously I encountered error like:

imrad/src/binding.h:184:24: error: member access into incomplete type 'const UIContext'
                const auto* var = ctx.codeGen->GetVar(str);
                                     ^
imrad/src/binding.h:11:8: note: forward declaration of 'UIContext'
struct UIContext;
       ^
1 error generated.

In file included from imrad/src/binding.cpp:1:
imrad/src/binding.h:185:32: error: member access into incomplete type 'CppGen'
                const auto* var = ctx.codeGen->GetVar(str);
                                             ^
imrad/src/ui_context.h:13:7: note: forward declaration of 'CppGen'
class CppGen;
      ^
Compiler ID:
Apple clang version 15.0.0 (clang-1500.3.9.4)
Target: arm64-apple-darwin23.4.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin

I therefore made my decision fixing it and it works. As this is the only ImGui editor I can easily compile in MacOS I really hope this help someone as well.

**Comment by mzccn** [comment]
@tpecholt @Super-Genius @tope99

**Comment by tpecholt** [comment]
Thanks for fixing the dependency issue

**Pull request merged**